### PR TITLE
Rework how arguments are passed to kernel launcher, refactor common code

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR mainly tries to make how we pass arguments through GPU kernels less cryptic. While there, also refactors common kernel launchers from CUDA and HIP driver layers to the top runtime layer.

### Implementation Overview:
- Introduces `kernel_cfg` type in the top runtime layer.
  - This is a struct that stores kernel parameters, the stream we're supposed to use, line/file no information etc.
  - It currently doesn't store the grid configuration or the kernel itself, but I think we can make that happen, and probably should.
- The main interface for this `kernel_cfg` is through:
  - `chpl_gpu_init_kernel_cfg`,
  - `chpl_gpu_deinit_kernel_cfg`,
  - `chpl_gpu_arg_offload`, (add an argument to the kernel that the runtime is supposed to offload)
  - `chpl_gpu_arg_pass` (similar, but the value runtime receives should be stored directly into the kernel parameters array)
- The compiler uses the following primitives to invoke these functions:
  - `PRIM_GPU_INIT_KERNEL_CFG`,
  - `PRIM_GPU_DEINIT_KERNEL_CFG`,
  - `PRIM_GPU_ARG`
  - All of these primitives are inserted during gpuTransforms
  - During codegen time, they are generated into correct runtime calls
- The compiler uses `GpuArgKind` enum to determine the codegen behavior for different type of passing arguments to the kernel. There was and still is three ways we do that. We are planning to expand that as we add implementation for intents in GPU kernels. And that's one of the main motivations of this PR.  

### More details:
- The `impl` layer's kernel launch implementation is very simple and only has `chpl_gpu_impl_kernel_launch` now.
- Most of the logic is now in the top layer and in the static helper `launch_kernel`. That function consumes `kernel_cfg` instances to launch kernels.
- Codegen for the actual launch primitive is much simpler as most of the argument passing logic is handled in gpuTransforms and codegen for `PRIM_GPU_ARG` and to some extent the runtime.

### Test status:
- [x] nvidia
- [x] amd
- [x] cpu spot-check
- [x] nvidia+gasnet spot-check

### CODE REVIEW

**Comment by e-kayrakli** [comment]
Here's a sample AST dump right after kernel generation. This is from jacobi:

```
if(2814884 '>='(2814883 'chpl_task_getRequestedSubloc') 0)
    {
      val ret[2814899]:DefaultRectangularArr(1,int(64),one,real(64),int(64))[1722389] "dead after last mention" "RVV" "temp"
      (2814901 'move' ret[2814899](2814903 '.v' A[2364110] _instance[1726983]))
      val coerce_tmp[2814906]:DefaultRectangularArr(1,int(64),one,real(64),int(64))[1722389] "dead after last mention" "coerce temp" "insert auto destroy" "temp"
      (2814908 'move' coerce_tmp[2814906] ret[2814899])
      val coerce_tmp[2814911]:_ddata(real(64))[1721472] "dead after last mention" "coerce temp" "insert auto destroy" "temp"
      (2814913 'move' coerce_tmp[2814911](2814915 '.v' coerce_tmp[2814906] shiftedData[1722366]))
      {
        val kernel_cfg[2816899]:chpl__c_void_ptr[134]
        (2816903 'move' kernel_cfg[2816899](2816901 'gpu init kernel cfg' 3))
        (2816905 'gpu arg' kernel_cfg[2816899] tmp_x0[2796227] 1)
        (2816909 'gpu arg' kernel_cfg[2816899] _ic__F1_high[2787630] 1)
        (2816913 'gpu arg' kernel_cfg[2816899] coerce_tmp[2814911] 1)
        val chpl_block_delta[2816917]:int(64)[10]
        val chpl_num_gpu_threads[2816919]:int(64)[10]
        (2816924 '=' chpl_block_delta[2816917](2816921 '-' _ic__F1_high[2787630] tmp_x0[2796227]))
        (2816929 '=' chpl_num_gpu_threads[2816919](2816926 '+' chpl_block_delta[2816917] 1))
        (2816931 'gpu kernel launch flat' chpl_gpu_kernel_jacobi_line_30_[2816785] chpl_num_gpu_threads[2816919] 512 kernel_cfg[2816899])
        (2816936 'gpu deinit kernel cfg' kernel_cfg[2816899])
      }
    }
    {
      CForLoop[2436197] order-independent
      {
```

**Review by stonea — APPROVED** [review]
I love this PR. The code is so much cleaner in the runtime and the fact that there's now an API to dictate how things are done on a parameter-by-parameter basis will make it a lot easier for gpuTransforms to modify the behavior as we add support for `foreach` intents.

File: compiler/include/misc.h [hunk 1]
```diff
@@ -96,6 +96,13 @@ enum class GpuCodegenType {
   GPU_CG_CPU,
 };
 
+enum GpuArgKind {
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/include/misc.h:101]: > That is: a GpuArgKind isn't a single value thats a choice between these three unique values in the enum, rather it's a 2-bit value where one bit dictates whether its offloaded or not and one bit dictates whether its passed in directly or by address.

Yeah, I imagine a bitmask can give us a better flexibility down the road. But while thinking about your comment, I realize that having 3 values might have also been confusing. Do you think https://github.com/chapel-lang/chapel/pull/24047/commits/de6dcaa9fde3b5b6e453503b8828604e48c768e0 improves the situation?

[the remark this answers is not in the document]
Comment (stonea) [reply compiler/include/misc.h:94]: Yep, I think that helps.

**Pull request merged**